cmd/k8s-operator/ingress-for-pg.go

```diff
@@ -0,0 +1,524 @@
+	// VIPService in this way won't be incorrect.
+	dnsNameBase, err := a.dnsNameBaseForVIPService(ctx)
+	if err != nil {
+		return fmt.Errorf("error determining DNS name base: %w", err)
+	}
+	dnsName := hostname + "." + dnsNameBase
+	svc, err := a.tsClient.getVIPServiceByName(ctx, hostname)
+	// TODO(irbekrm): here and when creating the VIPService, verify if the error is not terminal (and therefore
+	// should not be reconciled). For example, if the hostname is already a hostname of a Tailscale node, the GET
+	// here will fail.
+	if err != nil {
+		errResp := &tailscale.ErrResponse{}
+		if ok := errors.As(err, errResp); ok && errResp.Status != http.StatusNotFound {
+			logger.Infof("error getting VIPService %q: %v", hostname, err)
+			return fmt.Errorf("error getting VIPService %q: %w", hostname, err)
+		}
+	}
+	if svc != nil && !isVIPServiceForIngress(svc, ing) {
+		logger.Infof("VIPService %q for MagicDNS name %q  already exists, but is not owned by this Ingress. Please delete it manually and recreate this Ingress to proceed or create an Ingress for a different MagicDNS name", hostname, dnsName)
+		a.recorder.Event(ing, corev1.EventTypeWarning, "ConflictingVIPServiceExists", fmt.Sprintf("VIPService %q for MagicDNS name %q already exists, but is not owned by this Ingress. Please delete it manually to proceed or create an Ingress for a different MagicDNS name", hostname, dnsName))
+		return nil
+	}
+
+	// 3. Ensure that the serve config for the ProxyGroup contains the VIPService
+	cm, cfg, err := a.proxyGroupServeConfig(ctx, pg)
+	if err != nil {
+		return fmt.Errorf("error getting ingress serve config: %w", err)
+	}
+	ep := ipn.HostPort(fmt.Sprintf("%s:443", dnsName))
+	handlers, err := handlersForIngress(ctx, ing, a.Client, a.recorder, dnsName, logger)
+	if err != nil {
+		return fmt.Errorf("failed to get handlers for ingress: %w", err)
+	}
+	ingCfg := &ipn.ServiceConfig{
+		TCP: map[uint16]*ipn.TCPPortHandler{
+			443: {
+				HTTPS: true,
+			},
+		},
+		Web: map[ipn.HostPort]*ipn.WebServerConfig{
+			ep: {
+				Handlers: handlers,
+			},
+		},
+	}
+	var gotCfg *ipn.ServiceConfig
+	if cfg != nil && cfg.Services != nil {
+		gotCfg = cfg.Services[hostname]
+	}
+	if !reflect.DeepEqual(gotCfg, ingCfg) {
+		logger.Infof("Updating serve config")
+		mak.Set(&cfg.Services, hostname, ingCfg)
+		cfgBytes, err := json.Marshal(cfg)
+		if err != nil {
+			return fmt.Errorf("error marshaling serve config: %w", err)
+		}
+		mak.Set(&cm.BinaryData, serveConfigKey, cfgBytes)
+		if err := a.Update(ctx, cm); err != nil {
+			return fmt.Errorf("error updating serve config: %w", err)
+		}
+	}
+
+	// 4. Ensure that the VIPService exists and is up to date.
+	tags := a.defaultTags
+	if tstr, ok := ing.Annotations[AnnotationTags]; ok {
+		tags = strings.Split(tstr, ",")
+	}
+
+	vipSVc := &VIPService{
+		Name:    hostname,
+		Tags:    tags,
+		Comment: fmt.Sprintf(VIPSvcOwnerRef, ing.UID),
```

Comment:
In fact, I might make the comment bit contain a json string and utilize this for multi-cluster co-ordination
i.e 
```
{"tailscale.com/k8s-operator-ownerrefs": ["uid1","uid2","uid3"]}
```
or similar.

That of course wouldn't be the same as having an atomic 'delete-if-no-owner-ref' operation.
But in practice that _might _ be good enough (perhaps till we see how it works and can discuss whether we can add owner refs to the API) because I imagine that this reconcile loop will also run a healthcheck against the VIPService before assuming that the Ingress can be marked as ready - so would be more likely (although not entirely safe) that a racing DELETE by another operator would get caught.

But yeah, I would like to get this working as a prototype and then iterate on top.

Reply:
As discussed offline I think we should consider changing the comment field in the API to a `map[string]string` if we want to do this. I am fine with landing as-is while we iterate though.